File: lib/services/database_service.dart
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';
import 'dart:async';

import '../models/session.dart';
import '../models/achievement.dart';
import '../models/daily_stats.dart';
import '../models/user_preferences.dart';
import '../models/notification_settings.dart';
import '../models/biometric_data.dart';
import '../utils/constants.dart';

class DatabaseService {
  static final DatabaseService _instance = DatabaseService._internal();
  static Database? _database;

  factory DatabaseService() => _instance;

  DatabaseService._internal();

  static DatabaseService get instance => _instance;

  Future<Database> get database async {
    if (_database != null) return _database!;
    _database = await _initDatabase();
    return _database!;
  }

  Future<Database> _initDatabase() async {
    final directory = await getApplicationDocumentsDirectory();
    final path = join(directory.path, DbConstants.databaseName);

    return await openDatabase(
      path,
      version: DbConstants.databaseVersion,
      onCreate: _onCreate,
      onUpgrade: _onUpgrade,
      onConfigure: _onConfigure,
    );
  }

  Future<void> _onConfigure(Database db) async {
    await db.execute('PRAGMA foreign_keys = ON');
  }

  Future<void> _onCreate(Database db, int version) async {
    await db.execute('''
      CREATE TABLE ${DbConstants.sessionsTable} (
        id TEXT PRIMARY KEY,
        exercise_id TEXT NOT NULL,
        emotion_id TEXT NOT NULL,
        start_time TEXT NOT NULL,
        end_time TEXT,
        duration INTEGER NOT NULL DEFAULT 0,
        stress_before INTEGER,
        stress_after INTEGER,
        heart_rate_before REAL,
        heart_rate_after REAL,
        completed INTEGER NOT NULL DEFAULT 0,
        created_at TEXT NOT NULL DEFAULT (datetime('now'))
      )
    ''');

    await db.execute('''
      CREATE TABLE ${DbConstants.achievementsTable} (
        id TEXT PRIMARY KEY,
        title TEXT NOT NULL,
        description TEXT NOT NULL,
        icon TEXT NOT NULL,
        type TEXT NOT NULL,
        condition_key TEXT NOT NULL,
        current_progress INTEGER NOT NULL DEFAULT 0,
        target INTEGER NOT NULL DEFAULT 1,
        unlocked INTEGER NOT NULL DEFAULT 0,
        unlocked_at TEXT,
        created_at TEXT NOT NULL DEFAULT (datetime('now'))
      )
    ''');

    await db.execute('''
      CREATE TABLE ${DbConstants.dailyStatsTable} (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        date TEXT NOT NULL UNIQUE,
        session_count INTEGER NOT NULL DEFAULT 0,
        total_duration INTEGER NOT NULL DEFAULT 0,
        avg_stress REAL,
        sleep_quality REAL,
        exercise_types_completed TEXT,
        created_at TEXT NOT NULL DEFAULT (datetime('now')),
        updated_at TEXT NOT NULL DEFAULT (datetime('now'))
      )
    ''');

    await db.execute('''
      CREATE TABLE ${DbConstants.userPreferencesTable} (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        key TEXT NOT NULL UNIQUE,
        value TEXT NOT NULL,
        updated_at TEXT NOT NULL DEFAULT (datetime('now'))
      )
    ''');

    await db.execute('''
      CREATE TABLE ${DbConstants.notificationSettingsTable} (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        mode TEXT NOT NULL DEFAULT 'manual',
        scheduled_times TEXT,
        smart_enabled INTEGER NOT NULL DEFAULT 0,
        smart_threshold INTEGER NOT NULL DEFAULT 70,
        updated_at TEXT NOT NULL DEFAULT (datetime('now'))
      )
    ''');

    await db.execute('''
      CREATE TABLE ${DbConstants.biometricDataTable} (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        timestamp TEXT NOT NULL,
        heart_rate REAL,
        hrv REAL,
        sleep_quality REAL,
        activity_level REAL,
        respiratory_rate REAL,
        stress_level INTEGER,
        created_at TEXT NOT NULL DEFAULT (datetime('now'))
      )
    ''');

    await db.execute('''
      CREATE INDEX idx_sessions_start_time ON ${DbConstants.sessionsTable} (start_time)
    ''');

    await db.execute('''
      CREATE INDEX idx_sessions_emotion ON ${DbConstants.sessionsTable} (emotion_id)
    ''');

    await db.execute('''
      CREATE INDEX idx_daily_stats_date ON ${DbConstants.dailyStatsTable} (date)
    ''');

    await db.execute('''
      CREATE INDEX idx_biometric_timestamp ON ${DbConstants.biometricDataTable} (timestamp)
    ''');

    await _insertDefaultNotificationSettings(db);
  }

  Future<void> _onUpgrade(Database db, int oldVersion, int newVersion) async {
    if (oldVersion < 2) {
      await db.execute('''
        ALTER TABLE ${DbConstants.sessionsTable} 
        ADD COLUMN created_at TEXT NOT NULL DEFAULT (datetime('now'))
      ''');
    }
    if (oldVersion < 3) {
      await db.execute('''
        CREATE TABLE IF NOT EXISTS ${DbConstants.biometricDataTable} (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          timestamp TEXT NOT NULL,
          heart_rate REAL,
          hrv REAL,
          sleep_quality REAL,
          activity_level REAL,
          respiratory_rate REAL,
          stress_level INTEGER,
          created_at TEXT NOT NULL DEFAULT (datetime('now'))
        )
      ''');
    }
  }

  Future<void> _insertDefaultNotificationSettings(Database db) async {
    await db.insert(
      DbConstants.notificationSettingsTable,
      {
        'mode': 'manual',
        'scheduled_times': '[]',
        'smart_enabled': 0,
        'smart_threshold': 70,
        'updated_at': DateTime.now().toIso8601String(),
      },
    );
  }

  // ==================== Sessions CRUD ====================

  Future<String> insertSession(Session session) async {
    final db = await database;
    await db.insert(
      DbConstants.sessionsTable,
      session.toMap(),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
    return session.id;
  }

  Future<Session?> getSession(String id) async {
    final db = await database;
    final maps = await db.query(
      DbConstants.sessionsTable,
      where: 'id = ?',
      whereArgs: [id],
    );
    if (maps.isEmpty) return null;
    return Session.fromMap(maps.first);
  }

  Future<List<Session>> getAllSessions({int? limit, int? offset}) async {
    final db = await database;
    final maps = await db.query(
      DbConstants.sessionsTable,
      orderBy: 'start_time DESC',
      limit: limit,
      offset: offset,
    );
    return maps.map((map) => Session.fromMap(map)).toList();
  }

  Future<List<Session>> getSessionsByDateRange(
    DateTime start,
    DateTime end,
  ) async {
    final db = await database;
    final maps = await db.query(
      DbConstants.sessionsTable,
      where: 'start_time >= ? AND start_time <= ?',
      whereArgs: [start.toIso8601String(), end.toIso8601String()],
      orderBy: 'start_time DESC',
    );
    return maps.map((map) => Session.fromMap(map)).toList();
  }

  Future<List<Session>> getSessionsByEmotion(String emotionId) async {
    final db = await database;
    final maps = await db.query(
      DbConstants.sessionsTable,
      where: 'emotion_id = ?',
      whereArgs: [emotionId],
      orderBy: 'start_time DESC',
    );
    return maps.map((map) => Session.fromMap(map)).toList();
  }

  Future<List<Session>> getCompletedSessions({int? limit}) async {
    final db = await database;
    final maps = await db.query(
      DbConstants.sessionsTable,
      where: 'completed = 1',
      orderBy: 'start_time DESC',
      limit: limit,
    );
    return maps.map((map) => Session.fromMap(map)).toList();
  }

  Future<Session?> getLastSession() async {
    final db = await database;
    final maps = await db.query(
      DbConstants.sessionsTable,
      where: 'completed = 1',
      orderBy: 'start_time DESC',
      limit: 1,
    );
    if (maps.isEmpty) return null;
    return Session.fromMap(maps.first);
  }

  Future<int> updateSession(Session session) async {
    final db = await database;
    return await db.update(
      DbConstants.sessionsTable,
      session.toMap(),
      where: 'id = ?',
      whereArgs: [session.id],
    );
  }

  Future<int> deleteSession(String id) async {
    final db = await database;
    return await db.delete(
      DbConstants.sessionsTable,
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  Future<int> getSessionCount() async {
    final db = await database;
    final result = await db.rawQuery(
      'SELECT COUNT(*) as count FROM ${DbConstants.sessionsTable} WHERE completed = 1',
    );
    return Sqflite.firstIntValue(result) ?? 0;
  }

  Future<int> getTotalSessionDuration() async {
    final db = await database;
    final result = await db.rawQuery(
      'SELECT COALESCE(SUM(duration), 0) as total FROM ${DbConstants.sessionsTable} WHERE completed = 1',
    );
    return Sqflite.firstIntValue(result) ?? 0;
  }

  Future<int> getCurrentStreak() async {
    final db = await database;
    final results = await db.rawQuery('''
      SELECT DISTINCT date(start_time) as session_date
      FROM ${DbConstants.sessionsTable}
      WHERE completed = 1
      ORDER BY session_date DESC
    ''');

    if (results.isEmpty) return 0;

    int streak = 0;
    DateTime expectedDate = DateTime.now();
    expectedDate = DateTime(expectedDate.year, expectedDate.month, expectedDate.day);

    for (final row in results) {
      final sessionDate = DateTime.parse(row['session_date'] as String);
      final normalizedSessionDate = DateTime(
        sessionDate.year,
        sessionDate.month,
        sessionDate.day,
      );

      if (normalizedSessionDate == expectedDate) {
        streak++;
        expectedDate = expectedDate.subtract(const Duration(days: 1));
      } else if (normalizedSessionDate ==
          expectedDate.subtract(const Duration(days: 1))) {
        expectedDate = normalizedSessionDate;
        streak++;
        expectedDate = expectedDate.subtract(const Duration(days: 1));
      } else {
        break;
      }
    }

    return streak;
  }

  Future<int> getSessionCountByExerciseType(String exerciseType) async {
    final db = await database;
    final result = await db.rawQuery(
      '''
      SELECT COUNT(*) as count FROM ${DbConstants.sessionsTable} 
      WHERE completed = 1 AND exercise_id LIKE ?
      ''',
      ['%$exerciseType%'],
    );
    return Sqflite.firstIntValue(result) ?? 0;
  }

  // ==================== Achievements CRUD ====================

  Future<String> insertAchievement(Achievement achievement) async {
    final db = await database;
    await db.insert(
      DbConstants.achievementsTable,
      achievement.toMap(),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
    return achievement.id;
  }

  Future<Achievement?> getAchievement(String id) async {
    final db = await database;
    final maps = await db.query(
      DbConstants.achievementsTable,
      where: 'id = ?',
      whereArgs: [id],
    );
    if (maps.isEmpty) return null;
    return Achievement.fromMap(maps.first);
  }

  Future<List<Achievement>> getAllAchievements() async {
    final db = await database;
    final maps = await db.query(
      DbConstants.achievementsTable,
      orderBy: 'unlocked DESC, current_progress DESC',
    );
    return maps.map((map) => Achievement.fromMap(map)).toList();
  }

  Future<List<Achievement>> getUnlockedAchievements() async {
    final db = await database;
    final maps = await db.query(
      DbConstants.achievementsTable,
      where: 'unlocked = 1',
      orderBy: 'unlocked_at DESC',
    );
    return maps.map((map) => Achievement.fromMap(map)).toList();
  }

  Future<List<Achievement>> getLockedAchievements() async {
    final db = await database;
    final maps = await db.query(
      DbConstants.achievementsTable,
      where: 'unlocked = 0',
      orderBy: 'current_progress DESC',
    );
    return maps.map((map) => Achievement.fromMap(map)).toList();
  }

  Future<int> updateAchievement(Achievement achievement) async {
    final db = await database;
    return await db.update(
      DbConstants.achievementsTable,
      achievement.toMap(),
      where: 'id = ?',
      whereArgs: [achievement.id],
    );
  }

  Future<int> updateAchievementProgress(
    String id,
    int progress, {
    bool unlock = false,
  }) async {
    final db = await database;
    final Map<String, dynamic> values = {
      'current_progress': progress,
    };
    if (unlock) {
      values['unlocked'] = 1;
      values['unlocked_at'] = DateTime.now().toIso8601String();
    }
    return await db.update(
      DbConstants.achievementsTable,
      values,
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  Future<int> deleteAchievement(String id) async {
    final db = await database;
    return await db.delete(
      DbConstants.achievementsTable,
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  Future<void> insertAchievements(List<Achievement> achievements) async {
    final db = await database;
    final batch = db.batch();
    for (final achievement in achievements) {
      batch.insert(
        DbConstants.achievementsTable,
        achievement.toMap(),
        conflictAlgorithm: ConflictAlgorithm.ignore,
      );
    }
    await batch.commit(noResult: true);
  }

  // ==================== Daily Stats CRUD ====================

  Future<int> insertDailyStats(DailyStats stats) async {
    final db = await database;
    return await db.insert(
      DbConstants.dailyStatsTable,
      stats.toMap(),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  Future<DailyStats?> getDailyStats(String date) async {
    final db = await database;
    final maps = await db.query(
      DbConstants.dailyStatsTable,
      where: 'date = ?',
      whereArgs: [date],
    );
    if (maps.isEmpty) return null;
    return DailyStats.fromMap(maps.first);
  }

  Future<DailyStats?> getDailyStatsForDate(DateTime date) async {
    final dateStr =
        '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(

File: src/services/DatabaseService.ts
import * as SQLite from 'expo-sqlite';
import { Session } from '../models/Session';
import { Achievement } from '../models/Achievement';
import { Settings } from '../models/Settings';
import { UserStats } from '../models/User';

export class DatabaseService {
  private static instance: DatabaseService;
  private db: SQLite.SQLiteDatabase | null = null;

  private constructor() {}

  static getInstance(): DatabaseService {
    if (!DatabaseService.instance) {
      DatabaseService.instance = new DatabaseService();
    }
    return DatabaseService.instance;
  }

  async initialize(): Promise<void> {
    try {
      this.db = await SQLite.openDatabaseAsync('mindflow.db');
      await this.createTables();
      await this.initializeDefaultSettings();
      await this.initializeUserStats();
      await this.initializeAchievements();
    } catch (error) {
      console.error('Database initialization error:', error);
      throw error;
    }
  }

  private async createTables(): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    const tables = [
      `CREATE TABLE IF NOT EXISTS sessions (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        task_name TEXT,
        duration INTEGER NOT NULL,
        completed BOOLEAN NOT NULL DEFAULT 0,
        paused_count INTEGER DEFAULT 0,
        started_at TEXT NOT NULL,
        completed_at TEXT,
        created_at TEXT DEFAULT CURRENT_TIMESTAMP
      )`,
      
      `CREATE TABLE IF NOT EXISTS breaks (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        session_id INTEGER,
        duration INTEGER NOT NULL,
        skipped BOOLEAN DEFAULT 0,
        started_at TEXT NOT NULL,
        completed_at TEXT,
        created_at TEXT DEFAULT CURRENT_TIMESTAMP,
        FOREIGN KEY(session_id) REFERENCES sessions(id)
      )`,
      
      `CREATE TABLE IF NOT EXISTS achievements (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        type TEXT NOT NULL UNIQUE,
        title TEXT NOT NULL,
        description TEXT,
        unlocked BOOLEAN DEFAULT 0,
        unlocked_at TEXT,
        created_at TEXT DEFAULT CURRENT_TIMESTAMP
      )`,
      
      `CREATE TABLE IF NOT EXISTS settings (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        session_duration INTEGER DEFAULT 15,
        break_duration INTEGER DEFAULT 5,
        daily_goal INTEGER DEFAULT 5,
        sound_enabled BOOLEAN DEFAULT 1,
        vibration_enabled BOOLEAN DEFAULT 1,
        notifications_blocked BOOLEAN DEFAULT 1,
        work_start_time TEXT DEFAULT '09:00',
        work_end_time TEXT DEFAULT '17:00',
        updated_at TEXT DEFAULT CURRENT_TIMESTAMP
      )`,
      
      `CREATE TABLE IF NOT EXISTS user_stats (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        total_sessions INTEGER DEFAULT 0,
        total_focus_time INTEGER DEFAULT 0,
        total_breaks INTEGER DEFAULT 0,
        current_streak INTEGER DEFAULT 0,
        best_streak INTEGER DEFAULT 0,
        level INTEGER DEFAULT 1,
        stars INTEGER DEFAULT 0,
        updated_at TEXT DEFAULT CURRENT_TIMESTAMP
      )`
    ];

    for (const table of tables) {
      await this.db.execAsync(table);
    }
  }

  private async initializeDefaultSettings(): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    const result = await this.db.getFirstAsync<{ count: number }>(
      'SELECT COUNT(*) as count FROM settings'
    );

    if (result && result.count === 0) {
      await this.db.runAsync(
        `INSERT INTO settings (session_duration, break_duration, daily_goal, sound_enabled, vibration_enabled, notifications_blocked)
         VALUES (?, ?, ?, ?, ?, ?)`,
        [15, 5, 5, 1, 1, 1]
      );
    }
  }

  private async initializeUserStats(): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    const result = await this.db.getFirstAsync<{ count: number }>(
      'SELECT COUNT(*) as count FROM user_stats'
    );

    if (result && result.count === 0) {
      await this.db.runAsync(
        `INSERT INTO user_stats (total_sessions, total_focus_time, total_breaks, current_streak, best_streak, level, stars)
         VALUES (?, ?, ?, ?, ?, ?, ?)`,
        [0, 0, 0, 0, 0, 1, 0]
      );
    }
  }

  private async initializeAchievements(): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    const achievements = [
      { type: 'first_session', title: 'Первая сессия', description: 'Завершите первую сессию фокуса' },
      { type: 'five_sessions', title: 'Пять сессий', description: 'Завершите 5 сессий' },
      { type: 'ten_sessions', title: 'Десять сессий', description: 'Завершите 10 сессий' },
      { type: 'fifty_sessions', title: 'Пятьдесят сессий', description: 'Завершите 50 сессий' },
      { type: 'one_hour', title: 'Час фокуса', description: 'Накопите 1 час времени фокуса' },
      { type: 'five_hours', title: 'Пять часов', description: 'Накопите 5 часов времени фокуса' },
      { type: 'daily_goal', title: 'Дневная цель', description: 'Достигните дневной цели' },
      { type: 'week_streak', title: 'Неделя подряд', description: 'Занимайтесь 7 дней подряд' },
      { type: 'month_streak', title: 'Месяц подряд', description: 'Занимайтесь 30 дней подряд' },
      { type: 'no_pause', title: 'Без пауз', description: 'Завершите сессию без пауз' }
    ];

    for (const achievement of achievements) {
      try {
        await this.db.runAsync(
          `INSERT OR IGNORE INTO achievements (type, title, description, unlocked)
           VALUES (?, ?, ?, ?)`,
          [achievement.type, achievement.title, achievement.description, 0]
        );
      } catch (error) {
        console.error(`Error inserting achievement ${achievement.type}:`, error);
      }
    }
  }

  // Session methods
  async createSession(taskName: string, duration: number): Promise<number> {
    if (!this.db) throw new Error('Database not initialized');

    const result = await this.db.runAsync(
      `INSERT INTO sessions (task_name, duration, started_at)
       VALUES (?, ?, ?)`,
      [taskName || 'Сессия фокуса', duration, new Date().toISOString()]
    );

    return result.lastInsertRowId as number;
  }

  async completeSession(sessionId: number): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    await this.db.runAsync(
      `UPDATE sessions SET completed = 1, completed_at = ?
       WHERE id = ?`,
      [new Date().toISOString(), sessionId]
    );
  }

  async updateSessionPauseCount(sessionId: number): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    await this.db.runAsync(
      `UPDATE sessions SET paused_count = paused_count + 1
       WHERE id = ?`,
      [sessionId]
    );
  }

  async getSession(sessionId: number): Promise<Session | null> {
    if (!this.db) throw new Error('Database not initialized');

    const session = await this.db.getFirstAsync<Session>(
      'SELECT * FROM sessions WHERE id = ?',
      [sessionId]
    );

    return session || null;
  }

  async getAllSessions(): Promise<Session[]> {
    if (!this.db) throw new Error('Database not initialized');

    const sessions = await this.db.getAllAsync<Session>(
      'SELECT * FROM sessions ORDER BY created_at DESC'
    );

    return sessions || [];
  }

  async getSessionsByDateRange(startDate: string, endDate: string): Promise<Session[]> {
    if (!this.db) throw new Error('Database not initialized');

    const sessions = await this.db.getAllAsync<Session>(
      `SELECT * FROM sessions 
       WHERE created_at >= ? AND created_at <= ?
       ORDER BY created_at DESC`,
      [startDate, endDate]
    );

    return sessions || [];
  }

  async getTodaySessions(): Promise<Session[]> {
    if (!this.db) throw new Error('Database not initialized');

    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const tomorrow = new Date(today);
    tomorrow.setDate(tomorrow.getDate() + 1);

    const sessions = await this.db.getAllAsync<Session>(
      `SELECT * FROM sessions 
       WHERE created_at >= ? AND created_at < ?
       ORDER BY created_at DESC`,
      [today.toISOString(), tomorrow.toISOString()]
    );

    return sessions || [];
  }

  async getCompletedSessionsCount(): Promise<number> {
    if (!this.db) throw new Error('Database not initialized');

    const result = await this.db.getFirstAsync<{ count: number }>(
      'SELECT COUNT(*) as count FROM sessions WHERE completed = 1'
    );

    return result?.count || 0;
  }

  async getTotalFocusTime(): Promise<number> {
    if (!this.db) throw new Error('Database not initialized');

    const result = await this.db.getFirstAsync<{ total: number }>(
      'SELECT SUM(duration) as total FROM sessions WHERE completed = 1'
    );

    return result?.total || 0;
  }

  // Break methods
  async createBreak(sessionId: number, duration: number): Promise<number> {
    if (!this.db) throw new Error('Database not initialized');

    const result = await this.db.runAsync(
      `INSERT INTO breaks (session_id, duration, started_at)
       VALUES (?, ?, ?)`,
      [sessionId, duration, new Date().toISOString()]
    );

    return result.lastInsertRowId as number;
  }

  async completeBreak(breakId: number, skipped: boolean = false): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    await this.db.runAsync(
      `UPDATE breaks SET completed_at = ?, skipped = ?
       WHERE id = ?`,
      [new Date().toISOString(), skipped ? 1 : 0, breakId]
    );
  }

  async getBreaksBySessionId(sessionId: number): Promise<any[]> {
    if (!this.db) throw new Error('Database not initialized');

    const breaks = await this.db.getAllAsync(
      'SELECT * FROM breaks WHERE session_id = ? ORDER BY created_at DESC',
      [sessionId]
    );

    return breaks || [];
  }

  // Achievement methods
  async getAchievements(): Promise<Achievement[]> {
    if (!this.db) throw new Error('Database not initialized');

    const achievements = await this.db.getAllAsync<Achievement>(
      'SELECT * FROM achievements ORDER BY created_at ASC'
    );

    return achievements || [];
  }

  async unlockAchievement(type: string): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    await this.db.runAsync(
      `UPDATE achievements SET unlocked = 1, unlocked_at = ?
       WHERE type = ?`,
      [new Date().toISOString(), type]
    );
  }

  async getUnlockedAchievements(): Promise<Achievement[]> {
    if (!this.db) throw new Error('Database not initialized');

    const achievements = await this.db.getAllAsync<Achievement>(
      'SELECT * FROM achievements WHERE unlocked = 1 ORDER BY unlocked_at DESC'
    );

    return achievements || [];
  }

  async isAchievementUnlocked(type: string): Promise<boolean> {
    if (!this.db) throw new Error('Database not initialized');

    const result = await this.db.getFirstAsync<{ unlocked: number }>(
      'SELECT unlocked FROM achievements WHERE type = ?',
      [type]
    );

    return result?.unlocked === 1;
  }

  // Settings methods
  async getSettings(): Promise<Settings> {
    if (!this.db) throw new Error('Database not initialized');

    const settings = await this.db.getFirstAsync<Settings>(
      'SELECT * FROM settings LIMIT 1'
    );

    return settings || this.getDefaultSettings();
  }

  async updateSettings(settings: Partial<Settings>): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    const keys = Object.keys(settings).filter(key => key !== 'id');
    const values = keys.map(key => settings[key as keyof Settings]);
    
    const setClause = keys.map(key => `${key} = ?`).join(', ');
    const query = `UPDATE settings SET ${setClause}, updated_at = CURRENT_TIMESTAMP WHERE id = 1`;

    await this.db.runAsync(query, values);
  }

  private getDefaultSettings(): Settings {
    return {
      id: 1,
      session_duration: 15,
      break_duration: 5,
      daily_goal: 5,
      sound_enabled: true,
      vibration_enabled: true,
      notifications_blocked: true,
      work_start_time: '09:00',
      work_end_time: '17:00',
      updated_at: new Date().toISOString()
    };
  }

  // User stats methods
  async getUserStats(): Promise<UserStats> {
    if (!this.db) throw new Error('Database not initialized');

    const stats = await this.db.getFirstAsync<UserStats>(
      'SELECT * FROM user_stats LIMIT 1'
    );

    return stats || this.getDefaultUserStats();
  }

  async updateUserStats(stats: Partial<UserStats>): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    const keys = Object.keys(stats).filter(key => key !== 'id');
    const values = keys.map(key => stats[key as keyof UserStats]);
    
    const setClause = keys.map(key => `${key} = ?`).join(', ');
    const query = `UPDATE user_stats SET ${setClause}, updated_at = CURRENT_TIMESTAMP WHERE id = 1`;

    await this.db.runAsync(query, values);
  }

  async incrementTotalSessions(): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    await this.db.runAsync(
      'UPDATE user_stats SET total_sessions = total_sessions + 1, updated_at = CURRENT_TIMESTAMP WHERE id = 1'
    );
  }

  async addFocusTime(minutes: number): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    await this.db.runAsync(
      'UPDATE user_stats SET total_focus_time = total_focus_time + ?, updated_at = CURRENT_TIMESTAMP WHERE id = 1',
      [minutes]
    );
  }

  async incrementBreaks(): Promise<void> {
    if (!this.db) throw new Error('Database not initialized');

    await this.db.runAsync(
      'UPDATE user_stats SET total_breaks = total_breaks + 1, updated_at = CURRENT_TIMESTAMP WHERE id = 1'
    );
  }

  async updateStreak(currentStreak: number, bestStr